### ISSUE TITLE
Unexpected certificate.verify failure 

### ISSUE THREAD

**Issue by jcjones** [PeculiarVentures/PKI.js#126]
I've been using PKI.js within some automated tests and added a line to verify a self-signed certificate, roughly something like `attestationCert.verify()` which was expected to always be `true`. In 2-3% of cases, it wasn't.

I've captured a couple of the problematic certificates and included them below as `berBuf1` [[JS decode]](http://lapo.it/asn1js/#308201313081D9A003020102020500E79C7A2D300A06082A8648CE3D0403023021311F301D0603550403131646697265666F782055324620536F667420546F6B656E301E170D3137303933303139323233325A170D3137313030323139323233325A3021311F301D0603550403131646697265666F782055324620536F667420546F6B656E3059301306072A8648CE3D020106082A8648CE3D0301070342000485AEE0BF24CBBB3C23BA95BFE2703111E98A9A47E1BEA1EEB3748E52A915B1CA8671306823CBAD3C74E0487DC53E062BD256BAE3FBE7D867031B160F9C24D07D300A06082A8648CE3D04030203470030440220448791727E8B93053C54E997FA6750DE3510AD43A9E5FDB11FFA811021A2E447022000C18EC9BF1032A40302D9A46757EFE37E9207FD368E2EB8B57876C355275A5A) and `berBuf2` [[JS decode]](http://lapo.it/asn1js/#308201313081D9A003020102020500AB4D8C79300A06082A8648CE3D0403023021311F301D0603550403131646697265666F782055324620536F667420546F6B656E301E170D3137303932373136303033375A170D3137303932393136303033375A3021311F301D0603550403131646697265666F782055324620536F667420546F6B656E3059301306072A8648CE3D020106082A8648CE3D030107034200047E11731B2B502834F5A2E7618DC18BA914AA3890D56B59923A51B3E831C932964E0E032164647ADE0E93C82FF9AE16E89688CEDFFFA2B6C9DDBA34B962835150300A06082A8648CE3D0403020347003044022000E4692A3C2EF702899212313188931C2D16CD36FDF46E7C1244397A3B5E1CC50220561AEA29438D6DC102D9B87C936BDF22BF2B64273F711F32A3BE73E7A5F42E0E). 

Using my local copy of PKI.js-1.2.0 (release) and ASN1.js-1.1.0 (release), this test yields two `false` verifications:

```
let berBuf1 = new Uint8Array(window.atob("MIIBMTCB2aADAgECAgUA55x6LTAKBggqhkjOPQQDAjAhMR8wHQYDVQQDExZGaXJlZm94IFUyRiBTb2Z0IFRva2VuMB4XDTE3MDkzMDE5MjIzMloXDTE3MTAwMjE5MjIzMlowITEfMB0GA1UEAxMWRmlyZWZveCBVMkYgU29mdCBUb2tlbjBZMBMGByqGSM49AgEGCCqGSM49AwEHA0IABIWu4L8ky7s8I7qVv+JwMRHpippH4b6h7rN0jlKpFbHKhnEwaCPLrTx04Eh9xT4GK9JWuuP759hnAxsWD5wk0H0wCgYIKoZIzj0EAwIDRwAwRAIgRIeRcn6LkwU8VOmX+mdQ3jUQrUOp5f2xH/qBECGi5EcCIADBjsm/EDKkAwLZpGdX7+N+kgf9No4uuLV4dsNVJ1pa").split("").map((c) => { return c.charCodeAt(0); })).buffer;

let berBuf2 = new Uint8Array(window.atob("MIIBMTCB2aADAgECAgUAq02MeTAKBggqhkjOPQQDAjAhMR8wHQYDVQQDExZGaXJlZm94IFUyRiBTb2Z0IFRva2VuMB4XDTE3MDkyNzE2MDAzN1oXDTE3MDkyOTE2MDAzN1owITEfMB0GA1UEAxMWRmlyZWZveCBVMkYgU29mdCBUb2tlbjBZMBMGByqGSM49AgEGCCqGSM49AwEHA0IABH4RcxsrUCg09aLnYY3Bi6kUqjiQ1WtZkjpRs+gxyTKWTg4DIWRket4Ok8gv+a4W6JaIzt//orbJ3bo0uWKDUVAwCgYIKoZIzj0EAwIDRwAwRAIgAORpKjwu9wKJkhIxMYiTHC0WzTb99G58EkQ5ejteHMUCIFYa6ilDjW3BAtm4fJNr3yK/K2QnP3EfMqO+c+el9C4O").split("").map((c) => { return c.charCodeAt(0); })).buffer;

# Using PKI.js-1.2.0 and ASN1.js-1.1.0
new org.pkijs.simpl.CERT({ schema: org.pkijs.fromBER(berBuf1).result }).verify().then((r)=>{console.log("Verify berBuf1: ", r);}).catch((err)=>{console.log("Crash berBuf1: ", err)});

new org.pkijs.simpl.CERT({ schema: org.pkijs.fromBER(berBuf2).result }).verify().then((r)=>{console.log("Verify berBuf2: ", r);}).catch((err)=>{console.log("Crash berBuf2: ", err)});
````

Or, uusing what appears to be the v2 naming:
```
new Certificate({ schema: fromBER(berBuf1).result }).verify().then((r)=>{console.log("Verify berBuf1: ", r);}).catch((err)=>{console.log("Crash berBuf1: ", err)});

new Certificate({ schema: fromBER(berBuf2).result }).verify().then((r)=>{console.log("Verify berBuf2: ", r);}).catch((err)=>{console.log("Crash berBuf2: ", err)});
```

However, passing these certificates into OpenSSL shows them as OK:
```
echo "MIIBMTCB2aADAgECAgUAq02MeTAKBggqhkjOPQQDAjAhMR8wHQYDVQQDExZGaXJlZm94IFUyRiBTb2Z0IFRva2VuMB4XDTE3MDkyNzE2MDAzN1oXDTE3MDkyOTE2MDAzN1owITEfMB0GA1UEAxMWRmlyZWZveCBVMkYgU29mdCBUb2tlbjBZMBMGByqGSM49AgEGCCqGSM49AwEHA0IABH4RcxsrUCg09aLnYY3Bi6kUqjiQ1WtZkjpRs+gxyTKWTg4DIWRket4Ok8gv+a4W6JaIzt//orbJ3bo0uWKDUVAwCgYIKoZIzj0EAwIDRwAwRAIgAORpKjwu9wKJkhIxMYiTHC0WzTb99G58EkQ5ejteHMUCIFYa6ilDjW3BAtm4fJNr3yK/K2QnP3EfMqO+c+el9C4O"| base64 -D | openssl x509 -inform der > /tmp/ber2.crt ; openssl verify -verbose -CAfile /tmp/ber2.crt /tmp/ber2.crt
/tmp/ber2.crt: /CN=Firefox U2F Soft Token
error 10 at 0 depth lookup:certificate has expired
OK
```

```
echo "MIIBMTCB2aADAgECAgUA55x6LTAKBggqhkjOPQQDAjAhMR8wHQYDVQQDExZGaXJlZm94IFUyRiBTb2Z0IFRva2VuMB4XDTE3MDkzMDE5MjIzMloXDTE3MTAwMjE5MjIzMlowITEfMB0GA1UEAxMWRmlyZWZveCBVMkYgU29mdCBUb2tlbjBZMBMGByqGSM49AgEGCCqGSM49AwEHA0IABIWu4L8ky7s8I7qVv+JwMRHpippH4b6h7rN0jlKpFbHKhnEwaCPLrTx04Eh9xT4GK9JWuuP759hnAxsWD5wk0H0wCgYIKoZIzj0EAwIDRwAwRAIgRIeRcn6LkwU8VOmX+mdQ3jUQrUOp5f2xH/qBECGi5EcCIADBjsm/EDKkAwLZpGdX7+N+kgf9No4uuLV4dsNVJ1pa"| base64 -D | openssl x509 -inform der > /tmp/ber1.crt ; openssl verify -verbose -CAfile /tmp/ber1.crt /tmp/ber1.crt
/tmp/ber1.crt: /CN=Firefox U2F Soft Token
error 10 at 0 depth lookup:certificate has expired
OK
````

This may well be my error, or a truly problematic certificate -- but I haven't caught what might be wrong yet. 

Note: This is not a priority from my side, I'm treating it as a false-positive, but thought I'd record the test case for posterity!

**Comment by YuryStrozhevsky** [comment]
So, in both your certificates we have the same problem - strange encoding of ASN.1 integer DER values.

First of all I need to explain DER rules regarding integer encoding: integer must be encoded with minimal length of bytes and it should be non-negative. The "non-negative rule" is about "do not have first bit set to 1". 

In the first certificate signature looks like this:
```
224  10:   SEQUENCE {
226   8:     OBJECT IDENTIFIER ecdsaWithSHA256 (1 2 840 10045 4 3 2)
       :     }
236  71:   BIT STRING, encapsulates {
239  68:     SEQUENCE {
241  32:       INTEGER
       :         44 87 91 72 7E 8B 93 05 3C 54 E9 97 FA 67 50 DE
       :         35 10 AD 43 A9 E5 FD B1 1F FA 81 10 21 A2 E4 47
275  32:       INTEGER
       :         00 C1 8E C9 BF 10 32 A4 03 02 D9 A4 67 57 EF E3
       :         7E 92 07 FD 36 8E 2E B8 B5 78 76 C3 55 27 5A 5A
       :       }
       :     }
```
As we can see in the second integer we have "00" at the beginning and it is correct because then we have "C1" and this hex value has "1" in first bit. But the problem is that in fact this integer should be not 32 bytes long, but 33 bytes because you add the "00" byte in order to make the integer to be non-negative.

The same situation with second certificate:
```
224  10:   SEQUENCE {
226   8:     OBJECT IDENTIFIER ecdsaWithSHA256 (1 2 840 10045 4 3 2)
       :     }
236  71:   BIT STRING, encapsulates {
239  68:     SEQUENCE {
241  32:       INTEGER
       :         00 E4 69 2A 3C 2E F7 02 89 92 12 31 31 88 93 1C
       :         2D 16 CD 36 FD F4 6E 7C 12 44 39 7A 3B 5E 1C C5
275  32:       INTEGER
       :         56 1A EA 29 43 8D 6D C1 02 D9 B8 7C 93 6B DF 22
       :         BF 2B 64 27 3F 71 1F 32 A3 BE 73 E7 A5 F4 2E 0E
       :       }
       :     }
```
Here we have same situation, but with first integer: it should has length equal to 33 bytes, but has only 32.

I am not pretty sure why Windows, for example, does not say the certificate is incorrect, I will think about it. But I was trying to put additional "00" byte at the end of integers (probably Windows assuming this) but had the same result - signature can not be verified. So, the main problem here - why others consider it as correct signature? :)

**Comment by jcjones** [comment]
The `r` and `s` components in the sequence are variable length `INTEGER` types, they don't have to be 32-octets long exactly in the `prime256v1` case. A 31-octet long component is uncommon but not illegal.  (30-octet long components are also possible, but far more unlikely).

If you look at [how the NSS library unpacks ECDSA components](https://dxr.mozilla.org/mozilla-central/source/security/nss/lib/cryptohi/dsautil.c#70), you'll see that it tolerates shorter components.

**Comment by YuryStrozhevsky** [comment]
@jcjones Thanks for the hint, I will check the code.

**Comment by YuryStrozhevsky** [comment]
@jcjones I made a possible solution, please check. The solution is all about v2 PKI.js version code.

**Comment by jcjones** [comment]
I've taken a look, but the truth is that I don't know enough about your parser to really tell correctness.

I tried building the most recent revs of both ASN1.js and PKI.js to try it out, but I'm not experienced with Babel and don't think I built a functional library.

I'm guessing based on the general procedure that it's probably OK, but I'd recommend taking these certs and adding them to your unit tests just to watch for weird behavior in the future.

Thanks for working on this! I look forward to the next releases!
